### PR TITLE
Flax: Fix img2img and align with other pipeline

### PR DESCRIPTION
**Pull request by skirsten**
- Fixes for flax img2img
- Other misc changes
- Re-aligned the img2img pipe with the normal pipe (mostly copy paste)

### CODE REVIEW

File: src/diffusers/pipeline_flax_utils.py [hunk 1]
```diff
@@ -475,6 +475,49 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+        ...     FlaxStableDiffusionPipeline,
+        ...     FlaxStableDiffusionImg2ImgPipeline,
+        ... )
+
+        >>> text2img = FlaxStableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", revision="bf16", dtype=jnp.bfloat16)
+        >>> img2img = FlaxStableDiffusionImg2ImgPipeline(**text2img.components)
+        ```
+
+        Returns:
+            A dictionaly containing all the modules needed to initialize the pipeline.
```
Comment (pcuenca) [inline src/diffusers/pipeline_flax_utils.py:506]:
```suggestion
            A dictionary containing all the modules needed to initialize the pipeline.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 2]
```diff
@@ -400,13 +400,13 @@ def __call__(
     static_broadcasted_argnums=(0, 4, 5, 6),
 )
 def _p_generate(
-    pipe,
+    pipe,  # static
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:403]:
```suggestion
    pipe,
```

Can we please remove the line comments here?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 3]
```diff
@@ -400,13 +400,13 @@ def __call__(
...
     prompt_ids,
     params,
     prng_seed,
-    num_inference_steps,
-    height,
-    width,
+    num_inference_steps,  # static
+    height,  # static
+    width,  # static
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:409]:
```suggestion
    num_inference_steps,
    height,
    width,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 4]
```diff
@@ -287,108 +308,145 @@ def __call__(
...
+# Static argnums are pipe, start_timestep, num_inference_steps, height, width. A change would trigger recompilation.
+# Non-static args are (sharded) input tensors mapped over their first dimension (hence, `0`).
+@partial(
+    jax.pmap,
+    in_axes=(None, 0, 0, 0, 0, None, None, None, None, 0, 0, 0),
+    static_broadcasted_argnums=(0, 5, 6, 7, 8),
+)
 def _p_generate(
-    pipe,
+    pipe,  # static
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:425]:
```suggestion
    pipe,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 5]
```diff
@@ -287,108 +308,145 @@ def __call__(
...
     params,
     prng_seed,
-    strength,
-    num_inference_steps,
-    height,
-    width,
+    start_timestep,  # static
+    num_inference_steps,  # static
+    height,  # static
+    width,  # static
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:433]:
```suggestion
    start_timestep,
    num_inference_steps,
    height,
    width,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 6]
```diff
@@ -178,14 +179,19 @@ def _generate(
         prompt_ids: jnp.array,
         image: jnp.array,
         params: Union[Dict, FrozenDict],
-        prng_seed: jax.random.PRNGKey,
-        strength: float = 0.8,
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Ideally, we could make it so that `strength` can be a sharded array too (behaving exactly like `guidance_scale`) so we can perform multiple parallel requests with different strengths. This would be useful for batching requests, but we can address it in a future PR.

**Review by pcuenca — COMMENTED** [review]
Thanks a lot! This clearly improves the existing pipeline, I just left comments with a few suggestions and ideas for a potential improvement (make `strength` parallelizable too).

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 6 above)
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Initially I also wanted to do that, but AFIK its not possible (in an easy way):

The strength influences the `start_timestep` which is used in the `jax.lax.fori_loop` bounds. The bounds for the loop must be known at compile time (as with `num_inference_steps`) so AFAIK there is no way to have a different strength per shard in pmap.

The only way I could think of doing that is passing in a static `min_start_timestep` and using that as the loop bounds and then also passing in the dynamic sharded `start_timestep`. Then in the loop body, using a jax conditional discard / "skip" the body if `timestemp < start_timestep` but that would be overkill in my opinion.

Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Btw, the reason I did the `strength` -> `start_timestep` calculation outside the jit is to reduce the number of recompilations.
E.g. `strength(0.8) != strength(0.81)` but `start_timestep(0.8) == start_timestep(0.81)`

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 7]
```diff
@@ -164,12 +166,11 @@ def _run_safety_checker(self, images, safety_model_params, jit=False):
 
         return images, has_nsfw_concepts
 
-    def get_timestep_start(self, num_inference_steps, strength, scheduler_state):
+    def get_timestep_start(self, num_inference_steps, strength):
         # get the original timestep using init_timestep
-        offset = self.scheduler.config.get("steps_offset", 0)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: why is `offset` removed? We need to handle this for `DDIMScheduler`

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 6 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Agree with @skirsten, `strength` is used to determine the value of inference steps, and different `strength = different number of timesteps`, so it would be hard to parallelize it.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 8]
```diff
@@ -197,18 +203,32 @@ def _generate(
...
         context = jnp.concatenate([uncond_embeddings, text_embeddings])
 
+        latents_shape = (
+            batch_size,
+            self.unet.in_channels,
+            height // self.vae_scale_factor,
+            width // self.vae_scale_factor,
+        )
+        if noise is None:
+            noise = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:223]: any specific reason to hardcode the `dtype` to `jnp.float32`? think we should use `self.dtype` here as before for half-precision inference.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 9]
```diff
@@ -241,19 +261,19 @@ def loop_body(step, args):
...
+        latents = self.scheduler.add_noise(params["scheduler"], init_latents, noise, latent_timestep)
 
-        if debug:
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * params["scheduler"].init_noise_sigma
+
+        if DEBUG:
             # run with python for loop
-            for i in range(t_start, len(scheduler_state.timesteps)):
+            for i in range(start_timestep, num_inference_steps):
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: think it's safer to use `len(scheduler_state.timesteps)` that `num_inference_steps` because depending on the scheduler the could be some extra timesteps, for example if using `PNDMScheduler` with PRK stpes, it'll add some extra time timesteps

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 10]
```diff
@@ -178,14 +179,19 @@ def _generate(
...
+        height: Optional[int] = None,
+        width: Optional[int] = None,
         guidance_scale: float = 7.5,
-        debug: bool = False,
+        noise: Optional[jnp.array] = None,
+        neg_prompt_ids: jnp.array = None,
     ):
+        # 0. Default height and width to unet
+        height = height or self.unet.config.sample_size * self.vae_scale_factor
+        width = width or self.unet.config.sample_size * self.vae_scale_factor
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:193]: we compute the `height` and `width` in `__call__` as well, so here we could make it required.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for working on this! Left some comments

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 6 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Yes, good points @skirsten and @patil-suraj, let's leave it like this.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 7 above)
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: The offset cancels itself out. It was also removed from the other img2img pipelines in #1653

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 8 above)
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:223]: This was a copy paste from the normal pipeline. There are some other places where the `dtype` is hardcoded to `jnp.float32` related to latents. I just remember setting every occurrence to `self.dtype` and losing all details in the generated images.

I would prefer if all occurrences of `jnp.float32` could be removed at the same time. But I can also just revert this one change. Let me know what you think.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 9 above)
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: :thinking: I was not aware of that. Thanks for letting me know. In that case it would also have to be changed [here](https://github.com/huggingface/diffusers/blob/df2b548e893ccb8a888467c2508756680df22821/src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py#L271).

Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: Hmm, it looks to me like [here](https://github.com/huggingface/diffusers/blob/25e82ec47c40d5619e0772ed68f4082da4bd6457/src/diffusers/schedulers/scheduling_pndm_flax.py#L200) it is making sure that the length is always equal to `num_inference_steps` by dropping some plms timesteps.

Or am I getting that wrong?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 7 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: Oh I see, didn't notice that, thanks!

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 8 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: cc @pcuenca do we need noise in float32 in Jax?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 9 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Actually you are right, let's use `num_inference_steps` here. Because the timesteps are extended for 2nd order schedulers like Heun.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 8 above)
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: I investigated this a bit more and found the "problem". I asked the JAX maintainers on their thoughts here: https://github.com/google/jax/discussions/13798

Basically, to prevent losing detail in the image, the noise has to be generated in `float32` and then casted to `self.dtype`:
```py
noise = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32).astype(self.dtype)
```

Still I would prefer not to do that in this PR and instead create a new PR that fixes all of these occurrences at once.

**Review by patil-suraj — APPROVED** [review]
All looks good now, thank you for addressing the comments! The quality checks are failing, run `make style` and `make quality` , push and then it should be good to merge after green light from @pcuenca

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 8 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: That experiment was very cool and instructive, thanks a lot for taking the time to clarify the behaviour! I agree to deal with this in another PR.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 9 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Hmm, I'm not sure about this. Because the timesteps are extended for some schedulers, shouldn't we loop through the `timesteps` instead?

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Those schedulers are not part of flax yet no? (specifically the `HeunScheduler`),

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: No, they are not.

**Review by pcuenca — APPROVED** [review]

**Comment by skirsten** [comment]
The failing CI seems to be unrelated to these changes (doc_builder)

**Comment by pcuenca** [comment]
@skirsten This is what `make style` did for me: a47d1f07, you probably didn't install `doc-builder`.

Feel free to cherry-pick in your branch :)

**Comment by skirsten** [comment]
@pcuenca Thanks, I added you commit. Yes, I was missing some packages (and make did not resolve python to python3) so I just assumed that the problem was not in this branch :see_no_evil:

**Comment by pcuenca** [comment]
No worries!

**Comment by pcuenca** [comment]
@skirsten @patil-suraj is this ready to merge then?
